File: csharp/Basket/Basket.cs
namespace Basket {

	using System;
	using Products;
	using Warehouse;
	using System.Collections.Generic;
	using System.Linq;

	public class Basket {

		private IDictionary<string, BasketItem> items;
		
		public Basket() {
			items = new Dictionary<string, BasketItem>();
		}

		public IList<BasketItem> List() {
			var result = new List<BasketItem>();
			foreach (string sku in SortedSkus())
				result.Add(items[sku]);
			return result;
		}

		private IList<string> SortedSkus() {
			return items.Keys.OrderBy(k => k).ToList();
		}

		public void Add(string skuId, Catalogue catalogue, int numItems) {
			var sku = catalogue.Lookup(skuId);
			if (items.ContainsKey(skuId)) {
				var existing = items[skuId];
				items.Add(skuId, new BasketItem(skuId, existing.title, existing.price, existing.count + numItems));
			} else {
				items.Add(skuId, new BasketItem(skuId, sku.title, sku.price, numItems));
			}
		}

		public void Checkout(Warehouse warehouse) {
			if (items.Count == 0) {
				Console.WriteLine("Your basket is empty!");
				return;
			}
			double totalPrice = CurrentTotal() / 100.0;
			foreach (string sku in items.Keys)
				warehouse.Fulfill(sku, items[sku].count);
			items = new Dictionary<string, BasketItem>();
			Console.WriteLine("All items checked out. Total price £{0,5:F2}", totalPrice);
		}

		private int CurrentTotal() {
			var total = 0;
			foreach (BasketItem item in items.Values)
				total += item.count * item.price;
			if (total > 2000)
				total -= (total/10);
			return total;
		}

	}
}

File: python3/basket/basket.py
from . basket_item import BasketItem

class Basket:

    def __init__(self):
        self.items = {}

    def list(self):
        return [self.items[sku] for sku in self._sortedSkus()]

    def add(self, skuId, catalogue, numItems):
        sku = catalogue.lookup(skuId)
        if skuId in self.items:
            existing = self.items[skuId]
            self.items[skuId] = BasketItem(skuId, existing.title, existing.price, existing.count + numItems)
        else:
            self.items[skuId] = BasketItem(skuId, sku.title, sku.price, numItems)

    def checkout(self, warehouse):
        if not self.items:
            print("Your basket is empty!")
            return
        for sku in self.items.keys():
            warehouse.fulfill(sku, self.items[sku].count)
        self.items = {}
        print("All items checked out. Total price $%5.02f\n" % (self._currentTotal() / 100.0))

    def _sortedSkus(self):
        return sorted(self.items.keys())

    def _currentTotal(self):
        total = 0
        for item in self.items.values():
            total += item.count * item.price
        if total > 2000:
            total -= (total/10)
        return total

